src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala

```diff
@@ -526,12 +526,19 @@ abstract class BCodeBodyBuilder extends BCodeSkelBuilder {
             if (l.isPrimitive && r.isPrimitive)
               genConversion(l, r, cast)
             else if (l.isPrimitive) {
-              bc drop l
               if (cast) {
-                mnode.visitTypeInsn(asm.Opcodes.NEW, jlClassCastExceptionRef.internalName)
-                bc dup ObjectRef
-                emit(asm.Opcodes.ATHROW)
+                if (l == UNIT && r == srBoxedUnitRef) {
+                  /* special case: we can "cast" V to Lscala/runtime/BoxedUnit; by summoning the latter */
+                  mnode.instructions.add(backendUtils.getBoxedUnit)
```

Comment:
I think this should be handled in erasure, at the place where the cast is inserted, see where [`cast`](https://github.com/scala/scala/blob/9ad9297e35455d3e6a36f28250c8afbf9fff0071/src/compiler/scala/tools/nsc/transform/TypeAdaptingTransformer.scala#L102) is invoked.